### ISSUE TITLE
Stack overflow

### ISSUE THREAD

**Issue by BrianHicks** [elm-tools/parser#5]
I've found a particular combination of parsers that, when combined, overflow the stack.

```elm
import Char
import Html exposing (..)
import Parser exposing (..)


main =
  text <| toString <| run root "a/b/c"

root =
  zeroOrMore (element |. optional separator)

element =
  mapWithSource always (ignoreWhile Char.isLower)

optional p =
  oneOf [ p, succeed () ]

separator =
  symbol "/"
```

I can't seem to get it any smaller, unfortunately. The two important parts: `ignoreWhile Char.isLower` and `optional`. If either of those are removed from the execution path (by dropping `optional` or using `ignore 1 Char.isLower`), the parser works.

**Comment by evancz** [comment]
The problem is that `ignoreWhile` will succeed even when it consumes nothing. So the parser is successfully parsing `a/` then `b/` then `c` and then it is parsing "more" of nothing. That just keeps succeeding, leading to the crash.

It seems like this is not good default behavior, so it makes sense to fix this when the whole "`ignore` is confusing" problem is fixed.

**Comment by BrianHicks** [comment]
Is there a reasonable workaround in the meantime? The general form of my use case is "ignore this if it's present, but don't worry too much otherwise"

**Comment by evancz** [comment]
I think replacing `ignoreWhile Char.isLower` with `zeroOrMore (ignore 1 Char.isLower)` would do it.

If not, I'd check out [the implementation of `zeroOrMore`](https://github.com/elm-tools/parser/blob/master/src/Parser.elm#L453-L464) and use that as a model to write the behavior you are looking for.

**Comment by evancz** [comment]
This was fixed by the #6 fix actually! I now detect when `zeroOrMore` is infinitely consuming nothing, so the program above now produces:

```elm
Err { row = 1, col = 6, source = "a/b/c", problem = BadRepeat, context = [] }
```

**Comment by evancz** [comment]
Okay, the new version is out as 2.0.0. I fixed this issue, but I also simplified the API quite a bit.
